### PR TITLE
feat: convert storage to typescript

### PR DESCRIPTION
**Pull request by viceice**
This pr move the `lib/platform/git/storage.js` file to typescript.

*Changes*
* Add some external type declarations
* Rename file
* Add required type definitions (required by no-implicit-any)
* changed some typescript options to exclude default browser declarations (`lib.dom` and others, which are not available in node and are not required -> speed-up)

### CODE REVIEW

File: lib/types.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+declare namespace Renovate {
```
Comment (rarkins) [inline lib/types.d.ts]: Is this file necessary for this PR?

Comment (ViceIce) [reply lib/types.d.ts]: The typings: yes
The namespace: can be removed, will possilby collide with global types if removed. Some kind of best practise to separate own typings

Comment (rarkins) [reply lib/types.d.ts]: ok

**Comment by ViceIce** [comment]
rebased to resolve conficts

File: lib/platform/git/storage.ts [hunk 2]
```diff
@@ -1,15 +1,33 @@
-const fs = require('fs-extra');
-const { join } = require('path');
-const path = require('path');
-const URL = require('url');
-const Git = require('simple-git/promise');
-const convertHrtime = require('convert-hrtime');
+/// <reference path="../../../node_modules/simple-git/promise.d.ts" />
```
Comment (JamieMagee) [inline lib/platform/git/storage.ts]: I'm not sure this is the best practices anymore for including typings. I think you should define them in the `tsconfig.json` instead. I don't think this is really sustainable.

File: lib/platform/git/storage.ts [hunk 3]
```diff
@@ -1,15 +1,33 @@
...
+
+import fs from 'fs-extra';
+import { join } from 'path';
+import path from 'path';
+import URL from 'url';
+import Git from 'simple-git/promise';
+import convertHrtime from 'convert-hrtime';
+
+declare module 'fs-extra' {
+  export function exists(pathLike: string): Promise<boolean>;
```
Comment (JamieMagee) [inline lib/platform/git/storage.ts]: Is this necessary? Doesn't `fs-extra` have its own typings?

File: lib/platform/git/storage.ts (hunk 2 above)
Comment (ViceIce) [reply lib/platform/git/storage.ts]: I've removed that on my ts3 branch. Is no longer required.

File: lib/platform/git/storage.ts (hunk 3 above)
Comment (ViceIce) [reply lib/platform/git/storage.ts]: Yes, but this overload is missing

File: lib/platform/git/storage.ts (hunk 2 above)
Comment (ViceIce) [reply lib/platform/git/storage.ts]: removed

**Comment by ViceIce** [comment]
I'm ok with this, running fine against gitlab-ce and bitbucket-server.
When this is merged i'll rebase part2, don't want to make to much noise. 😄
Part3 is ready too, should i merge with part 2?

@JamieMagee Please have a review, so this can be merged to master

File: lib/platform/git/storage.ts [hunk 4]
```diff
@@ -152,17 +167,17 @@ class Storage {
...
     }
 
-    async function createBranch(branchName, sha) {
+    async function createBranch(branchName: string, sha: string) {
       logger.debug(`createBranch(${branchName})`);
       await git.reset('hard');
       await git.raw(['clean', '-fd']);
       await git.checkout(['-B', branchName, sha]);
-      await git.push(['origin', branchName, '--force']);
+      await git.push('origin', branchName, { '--force': true });
```
Comment (rarkins) [inline lib/platform/git/storage.ts]: This change is unrelated to TS, right?

File: lib/platform/git/storage.ts [hunk 5]
```diff
@@ -347,12 +362,10 @@ class Storage {
           }
         }
         await git.commit(message);
-        await git.push([
-          'origin',
-          `${branchName}:${branchName}`,
-          '--force',
-          '-u',
-        ]);
+        await git.push('origin', `${branchName}:${branchName}`, {
```
Comment (rarkins) [inline lib/platform/git/storage.ts]: Should we move these git changes to a smaller, separate PR to merge first?

File: lib/platform/git/storage.ts (hunk 4 above)
Comment (ViceIce) [reply lib/platform/git/storage.ts]: yes, should do the same, the array overload is missing in types, so i have to use the options overload

File: lib/platform/git/storage.ts (hunk 5 above)
Comment (ViceIce) [reply lib/platform/git/storage.ts]: ok, no problem. give me a few minutes 😄

Comment (ViceIce) [reply lib/platform/git/storage.ts]: done, see #3701

Comment (rarkins) [reply lib/platform/git/storage.ts]: Merged

**Comment by ViceIce** [comment]
Hopefully now everything is fine 😄

**Comment by ViceIce** [comment]
hopefully no more conflicts 🙈
